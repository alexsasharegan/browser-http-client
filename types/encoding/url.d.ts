export declare type Primitives = string | number | boolean | null;
export declare type PrimitiveArray = string[] | number[] | boolean[] | null[];
export declare type QueryObject = {
    [index: string]: Primitives | PrimitiveArray;
};
export declare function EncodeQuery(data: QueryObject): string;
