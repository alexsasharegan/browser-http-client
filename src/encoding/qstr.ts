export type Primitive = string | number | boolean | null;
export type PrimitiveArray = string[] | number[] | boolean[] | null[];
export type QueryObject = { [index: string]: Primitive | PrimitiveArray };

export function encode_query(data: QueryObject) {
  let encoded = [];
  let enc = encodeURIComponent;
  let val: any;
  let v: any;
  let k: string;

  for (k of Object.keys(data)) {
    val = data[k];
    if (typeof val == "function" || val === undefined) {
      // Skip QueryObject key iter for non encodable types.
      continue;
    }

    if (Array.isArray(val)) {
      for (v of val) {
        // Omit writing undefined value.
        if (v === undefined) {
          continue;
        }
        encoded.push(enc(k) + "[]=" + enc("" + v));
      }
      // Next QueryObject key.
      continue;
    }

    encoded.push(enc(k) + "=" + enc("" + val));
  }

  return encoded.join("&");
}
