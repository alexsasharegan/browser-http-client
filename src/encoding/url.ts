export type Primitives = string | number | boolean | null
export type PrimitiveArray = string[] | number[] | boolean[] | null[]
export type QueryObject = { [index: string]: Primitives | PrimitiveArray }

export function EncodeQuery(data: QueryObject) {
	let encoded = [],
		e = encodeURIComponent,
		val = undefined,
		v = undefined,
		k = undefined

	for (k of Object.keys(data)) {
		val = data[k]
		if (typeof val == "function" || val === undefined) {
			// Break QueryObject key iter for non encodable types.
			continue
		}

		if (Array.isArray(val)) {
			for (v of val) {
				if (v === undefined) {
					// Break inner array iter to omit writing undefined value.
					continue
				}
				encoded.push(e(k) + "[]=" + e(v.toString()))
			}
			// Continue to next QueryObject key.
			continue
		}

		encoded.push(e(k) + "=" + e(val.toString()))
	}

	return encoded.join("&")
}
