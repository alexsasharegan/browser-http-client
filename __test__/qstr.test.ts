import { encode_query } from "../src/encoding/qstr";

describe("encoding/qstr", async () => {
  it("should work with primitive types", async () => {
    let test = encode_query({
      string: "a string",
      number: 100,
      bool: false,
      empty: null,
    });

    // Snapshot was initially to grab value and manually check it.
    expect(test).toMatchSnapshot();
    // Snapshot value
    expect(test).toBe("string=a%20string&number=100&bool=false&empty=null");
  });

  it("should work with primitive array types", async () => {
    let test = encode_query({
      string: ["a string", "another string"],
      number: [100, 101],
      bool: [false, true],
      empty: [null, null],
      mixed: ["string", true, null, 100],
    });

    // Snapshot was initially to grab value and manually check it.
    expect(test).toMatchSnapshot();
    // Snapshot value
    expect(test).toBe(
      "string[]=a%20string&string[]=another%20string&number[]=100&number[]=101&bool[]=false&bool[]=true&empty[]=null&empty[]=null&mixed[]=string&mixed[]=true&mixed[]=null&mixed[]=100"
    );
  });

  it("should not include undefined values", async () => {
    let test = encode_query({ foo: undefined });
    expect(test).toBe("");
    expect(test).toMatchSnapshot();
  });

  it("should not include undefined values in an array", async () => {
    let test = encode_query({ foo: [undefined] });
    expect(test).toBe("");
    expect(test).toMatchSnapshot();
  });
});
