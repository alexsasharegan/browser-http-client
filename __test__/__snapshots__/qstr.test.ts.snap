// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`encoding/qstr should not include undefined values 1`] = `""`;

exports[`encoding/qstr should not include undefined values in an array 1`] = `""`;

exports[`encoding/qstr should work with primitive array types 1`] = `"string[]=a%20string&string[]=another%20string&number[]=100&number[]=101&bool[]=false&bool[]=true&empty[]=null&empty[]=null&mixed[]=string&mixed[]=true&mixed[]=null&mixed[]=100"`;

exports[`encoding/qstr should work with primitive types 1`] = `"string=a%20string&number=100&bool=false&empty=null"`;
